### ISSUE TITLE
Find similar observations using leaf node matching

### ISSUE THREAD

**Issue by mfdel** [dmlc/xgboost#11919]
I'm proposing a feature to find which training samples are "similar" to a prediction sample, based on whether they end up in the same leaf nodes across trees.

The idea is simple: if two observations consistently land in the same leaves across many trees, the model "sees" them as similar.

I already built a function to do this for my use case. I was doing a prediction to assess sales potential for some geographies and the model was predicting way too high for a zip code. I couldn't figure out why from the features or feature importances alone.

So I wrote a wrapper that checks: for this zip code, which training observations land in the same leaf nodes most often? Turns out a tiny zip code on the other side of the world matched 83% of the time. Looking at the features, they were strangely similar. It was an outlier and later I excluded it form the training.

I would never have found this by looking at Euclidean distance or raw feature values. The model knew these were similar. I just needed a way to ask it.

### This is how it works

1. Use `pred_leaf=True` to get leaf indices for each tree
2. For each tree, check if query and reference land in same leaf (boolean)
3. Average across trees → similarity score between 0 and 1 (I also weight trees by the variance of their leaf predictions, meaning that trees where all leaves predict ~same value aren't very discriminative.)

### What the API could look like

```python
# High-level
similar_idx, scores = model.find_similar(
    query=X_query,
    reference=X_train,
    k=5,
)

# Or lower-level on booster
query_leaves = booster.predict(DMatrix(X_query), pred_leaf=True)
ref_leaves = booster.predict(DMatrix(X_train), pred_leaf=True)
similarity = booster.compute_leaf_similarity(query_leaves, ref_leaves)
```

## Why this is useful

- **Debugging predictions**: "why is this prediction so high?" → find similar training samples and inspect them
- **Finding bad training data**: outliers in training can affect predictions in unexpected places
- **Explaining to stakeholders**: "this prediction is similar to these 5 historical cases" is easier to trust than a black box number

I haven't used it but I found out Random Forests have something similar to this, proximity matrices. Would be nice to have this in XGBoost.

## Questions

- Is this something that fits in XGBoost's scope, or better as a separate utility?
- Any concerns about scaling to large datasets? (the leaf prediction is fast, similarity is just broadcasting)
- Happy to put together a PR if there's interest

**Comment by trivialfis** [comment]
Hi, thank you for raising the issue. Yes , it's interesting and useful. Actually, I have been thinking about this but haven't been able to prioritize it yet. It would be great to have something more formal (mathmatical definitions that we can write down).

> Random Forests have something similar to this, proximity matrices.

I believe you are referring to something like this:

- https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm#prox
- https://ydf.readthedocs.io/en/stable/tutorial/counterfactual/

I haven't tried to extend it to GB trees yet. Seems plausible as you have suggested.

**Comment by trivialfis** [comment]
> Happy to put together a PR if there's interest

Please let us know if you are interested in this and we are happy to help. A great place to start would be describing the proximity a little bit more formally, for both exact solution and potentially the approximated solution (for memory efficiencies)

**Comment by mfdel** [comment]
Thanks for the feedback! Here's a more formal specification of the proximity/similarity metric.

## Formal Definition

### Notation

Let:
- $M$ be a trained XGBoost model with $T$ trees
- $\phi_t(x) \in \mathbb{Z}^+$ denote the leaf index that observation $x$ falls into in tree $t$
- $A = \{a_1, a_2, ..., a_m\}$ be the **query dataset**
- $B = \{b_1, b_2, ..., b_n\}$ be the **reference dataset**

### Definition 1: Basic Leaf Similarity

The **leaf similarity** between observation $a_i \in A$ and observation $b_j \in B$ is:

$$S(a_i, b_j) = \frac{1}{T} \sum_{t=1}^{T} \mathbb{1}[\phi_t(a_i) = \phi_t(b_j)]$$

Where $\mathbb{1}[\cdot]$ is the indicator function.

**Properties:**
- $S \in [0, 1]$
- $S = 1$ means both observations land in the same leaf in every tree
- $S = 0$ means they never share a leaf

This is identical to Breiman's Random Forest proximity matrix definition.

### Definition 2: Weighted Leaf Similarity

Not all trees are equally discriminative. A tree where all leaves predict similar values provides less signal.

Let $L_t = \{l_{t,1}, ..., l_{t,k_t}\}$ be the set of leaf predictions in tree $t$. Define **tree weight** as the variance of leaf predictions:

$$w_t = \text{Var}(L_t)$$

The **weighted leaf similarity** is:

$$S_w(a_i, b_j) = \frac{\sum_{t=1}^{T} w_t \cdot \mathbb{1}[\phi_t(a_i) = \phi_t(b_j)]}{\sum_{t=1}^{T} w_t}$$

**Intuition:** Trees where all leaves predict ~same value (low variance) contribute less to the similarity score.

### Top-k Similar Observations

Given query observation $a_i$ and reference dataset $B$, the **k most similar observations** are simply sorting by similarity and taking the top $k$:

$$\text{TopK}(a_i, B, k) = \text{argsort}_{b_j \in B}(S(a_i, b_j))[-k:]$$

---

### On Approximate Solutions

For the memory-efficient approximation, I don't have a fully worked-out solution yet beyond **sampling the reference dataset** — computing similarity against a random subset $B' \subset B$ rather than the full reference set.

More sophisticated approaches would need more research to formalize properly.

I could start with a PR implementing the **exact solution** (Definitions 1 & 2). We could add approximate methods in a follow-up if there's demand for very large datasets.

Would that be a reasonable approach?

**Comment by trivialfis** [comment]
Thank you for sharing. Skimming through the definition, the averaging makes sense for random forest. It might work with gradient boosting trees, but seems less intuitive as the result of each tree depends on the results of prior trees for gradient boosting. Is averaging a good way to handle that?

**Comment by mfdel** [comment]
I don't think the simple averaging (definition 1) is correct for boosting trees. I used it as an intro to explain definition 2. We need to apply some weights to trees when averaging since the effect of earlier trees are greater in the prediction.

**Comment by trivialfis** [comment]
> since the effect of earlier trees are greater in the prediction.

The second definition takes variance into account, but not the position of the tree?

**Comment by mfdel** [comment]
yes, correct. My objective was to differentiate trees with respect to their contribution to the model, how well the trees separate the samples. That's why I used variance. Variance of leaves decrease for later trees. The position of tree and variance is correlated but variance gives more information about the contribution of each tree to overall separation. 

We can definitely use alternative metrics. As I said, my objective was to differentiate trees with some metric. we can use total Gain achieved by each tree.

---
  
  Define **tree weight** as the total gain from all splits in tree $t$:
  
  $$w_t = \sum_{s \in \text{splits}(t)} \text{Gain}(s)$$
  
  Where $\text{Gain}(s)$ is the improvement in the objective function from split $s$.
  
  The **weighted leaf similarity** is:
  
  $$S_w(a_i, b_j) = \frac{\sum_{t=1}^{T} w_t \cdot \mathbb{1}[\phi_t(a_i) = \phi_t(b_j)]}{\sum_{t=1}^{T} w_t}$$

**Comment by trivialfis** [comment]
Total gain makes sense, maybe we can reuse the definition from feature importance, like hessian weight and gain weight.

**Comment by trivialfis** [comment]
In any case, I think we can pursue the feature and think more about generalization later.

**Comment by mfdel** [comment]
All right. here is the [PR](https://github.com/dmlc/xgboost/pull/11926).
